bot/cogs/verification.py

```diff
@@ -32,26 +59,325 @@
 
-    def __init__(self, bot: Bot):
+    # Cache last sent `REMINDER_MESSAGE` id
+    # RedisCache[str, discord.Message.id]
+    reminder_cache = RedisCache()
+
+    def __init__(self, bot: Bot) -> None:
+        """Start internal tasks."""
         self.bot = bot
 
+        self.update_unverified_members.start()
+        self.ping_unverified.start()
+
+    def cog_unload(self) -> None:
+        """
+        Cancel internal tasks.
+
+        This is necessary, as tasks are not automatically cancelled on cog unload.
+        """
+        self.update_unverified_members.cancel()
+        self.ping_unverified.cancel()
+
     @property
     def mod_log(self) -> ModLog:
         """Get currently loaded ModLog cog instance."""
         return self.bot.get_cog("ModLog")
 
+    # region: automatically update unverified users
+
+    async def _verify_kick(self, n_members: int) -> bool:
+        """
+        Determine whether `n_members` is a reasonable amount of members to kick.
+
+        First, `n_members` is checked against the size of the PyDis guild. If `n_members` are
+        more than `KICK_CONFIRMATION_THRESHOLD` of the guild, the operation must be confirmed
+        by staff in #core-dev. Otherwise, the operation is seen as safe.
+        """
+        log.debug(f"Checking whether {n_members} members are safe to kick")
+
+        await self.bot.wait_until_guild_available()  # Ensure cache is populated before we grab the guild
+        pydis = self.bot.get_guild(constants.Guild.id)
+
+        percentage = n_members / len(pydis.members)
+        if percentage < KICK_CONFIRMATION_THRESHOLD:
+            log.debug(f"Kicking {percentage:.2%} of the guild's population is seen as safe")
+            return True
+
+        # Since `n_members` is a suspiciously large number, we will ask for confirmation
+        log.debug("Amount of users is too large, requesting staff confirmation")
+
+        core_devs = pydis.get_channel(constants.Channels.dev_core)
+        confirmation_msg = await core_devs.send(
+            f"<@&{constants.Roles.core_developers}> Verification determined that `{n_members}` members should "
+            f"be kicked as they haven't verified in `{KICKED_AFTER}` days. This is `{percentage:.2%}` of the "
+            f"guild's population. Proceed?",
+            allowed_mentions=MENTION_CORE_DEVS,
+        )
+
+        options = (constants.Emojis.incident_actioned, constants.Emojis.incident_unactioned)
+        for option in options:
+            await confirmation_msg.add_reaction(option)
+
+        def check(reaction: discord.Reaction, user: discord.User) -> bool:
+            """Check whether `reaction` is a valid reaction to `confirmation_msg`."""
+            return (
+                reaction.message.id == confirmation_msg.id  # Reacted to `confirmation_msg`
+                and str(reaction.emoji) in options  # With one of `options`
+                and not user.bot  # By a human
+            )
```

Comment:
This will take a reaction from *any* user. Is this ok because the message is sent in a private channel, or should we add a sanity check for `user`'s roles in `pydis` too?

Reply:
Replaced the bot check with a role id check in f3c16f7.